Procfile: write logs to the tool home dir

david-caro opened the pull request:
This will require running with all the mounts:
```
toolforge job run \
    --mount=all \
    --schedule '0 19 */2 * *' \
    --command run-cron \
    --image tool-erex-yomi/tool-erex-yomi:latest \
    <job-name>
```

david-caro commented on Procfile:
```diff
@@ -1 +1 @@
-run-cron: ./erex_yomi.rb
+run-cron: bash -c "exec 1>$TOOL_DATA_DIR/out; exec 2>$TOOL_DATA_DIR/err; ./erex_yomi.rb"
```
You can change the name of the log files to whatever you prefer (here they are `out` and `err`)

abartov commented:
Thank you, this helped see the error. It was a gem loading error because the system Ruby (2.6) is too old. I had installed RVM to have a modern (userspace) Ruby available for the tool, and invoked it using the **run.sh** script, but didn't notice the Procfile here didn't use it.

I have changed it in 5cb5c1ec9971c8aaeed306559f458006e8f7a574, and hope it would work well tonight.

abartov commented:
@david-caro -- It has not been running these past few nights. 
```toolforge jobs logs erex-yomi``` shows nothing.

Is there something else I need to do after updating the Procfile?

david-caro commented:
> there something else I need to do after updating the Procfile?

Yes, every time you change the code (anything on this git repository), you have to rebuild the image, remember that the job does not use the code from the tool homedir, but from the git repository directly.

For a full refresh:
```
> toolforge build start https://github.com/abartov/erex_yomi
... wait for the build to finish
> toolforge jobs delete erex-yomi
> toolforge jobs run --schedule '0 19 */2 * *' --mount=all --command=run-cron --image=tool-erex-yomi/tool-erex-yomi:latest
```

Note that as I mentioned here https://github.com/abartov/erex_yomi/commit/5cb5c1ec9971c8aaeed306559f458006e8f7a574, the procfile should be using the file directly, not the script (that does not exist in the git repository), as in `run-cron: bash -c "exec 1>$TOOL_DATA_DIR/out; exec 2>$TOOL_DATA_DIR/err; ./erex_yomi.rb`

`toolforge jobs logs` will not show much if it shows anything, as all the output is being redirected to the `err` and `out` files on the tool home directory.

abartov commented:
Okay, thanks, but if that's the case, the issue is still an out-of-date Ruby. How can I get Ruby 3.x to be used? As mentioned above, I had installed RVM to have a fresh userspace Ruby available for the tool (and the run.sh script invokes it). The script can easily be checked into source control, but presumably that won't be enough, as the RVM environment would not be set up where it would be run.

I apologize for the clueless question: I just haven't had the time to learn the new runtime environment toolforge offers.

david-caro commented:
> Okay, thanks, but if that's the case, the issue is still an out-of-date Ruby. How can I get Ruby 3.x to be used? As mentioned above, I had installed RVM to have a fresh userspace Ruby available for the tool (and the run.sh script invokes it). The script can easily be checked into source control, but presumably that won't be enough, as the RVM environment would not be set up where it would be run.
> 
> I apologize for the clueless question: I just haven't had the time to learn the new runtime environment toolforge offers.

No problem at all, ask as much as you need :)

The environment provided by the image you build with the buildservice, is already using ruby 3.1:
```
dcaro@urcuchillay$ podman run --rm -ti --entrypoint launcher tools-harbor.wmcloud.org/tool-erex-yomi/tool-erex-yomi:latest bash

heroku@397b80e5fc56:/workspace$ ruby -v
ruby 3.1.3p185 (2022-11-24 revision 1a6b16756e) [x86_64-linux]
```

and should not need to use anything from your home directiory (code, rvm, etc.) to run the application, as the code is already inside the image itself (note that the path is `/workspace`, and that is *not* the tool home dir):
```
heroku@397b80e5fc56:/workspace$ ls -la
total 52
drwxrwxrwx. 1 heroku heroku  242 Jan  1  1980 .
dr-xr-xr-x. 1 root   root      6 Feb 14 11:17 ..
drwxrwxrwx. 1 heroku heroku   12 Jan  1  1980 .bundle
drwxrwxrwx. 1 heroku heroku  150 Jan  1  1980 .git
-rw-rw-rw-. 1 heroku heroku  159 Jan  1  1980 .gitignore
-rw-rw-rw-. 1 heroku heroku    6 Jan  1  1980 .ruby-version
-rw-rw-rw-. 1 heroku heroku   85 Jan  1  1980 Gemfile
-rw-rw-rw-. 1 heroku heroku 2386 Jan  1  1980 Gemfile.lock
-rw-rw-rw-. 1 heroku heroku 1211 Jan  1  1980 LICENSE
-rw-rw-rw-. 1 heroku heroku   83 Jan  1  1980 Procfile
-rw-rw-rw-. 1 heroku heroku  729 Jan  1  1980 README.md
-rw-rw-rw-. 1 heroku heroku   86 Jan  1  1980 TODO
-rwxrwxrwx. 1 heroku heroku 6082 Jan  1  1980 erex_yomi.rb
-rw-rw-rw-. 1 heroku heroku 5782 Jan  1  1980 fb_erex_yomi.rb
-rw-rw-rw-. 1 heroku heroku  392 Jan  1  1980 toolinfo.json
```
Can you try again using the `Procfile` without the `run.sh` script, and pasting the logs from it? (the `err/out files it will create`), I think that the issue might be something else.

abartov commented:
Thanks @david-caro, this worked and the tool ran!

However:

```
Hi!
Preparing featured article... - Title: חוק יסוד: נשיא המדינה - done!
Preparing Daily Quotation... done!
Preparing Today in History... done!
Preparing Today in Hebrew Calendar... done!
Preparing Recommended Picture... done!
Sending... sh: 1: /usr/sbin/sendmail: not found
done!
Bye!
```

Does the running environment not have a working sendmail(1)?

david-caro commented:
> Thanks @david-caro, this worked and the tool ran!
> 
> However:
> 
> ```
> Hi!
> Preparing featured article... - Title: חוק יסוד: נשיא המדינה - done!
> Preparing Daily Quotation... done!
> Preparing Today in History... done!
> Preparing Today in Hebrew Calendar... done!
> Preparing Recommended Picture... done!
> Sending... sh: 1: /usr/sbin/sendmail: not found
> done!
> Bye!
> ```
> 
> Does the running environment not have a working sendmail(1)?

It might not yes, you can install it by adding the appropriate package in the Aptfile, though you might need to configure it (try it first). I would recommend using ruby directly though (see https://www.magicbell.com/blog/how-to-send-email-using-ruby for tips).

abartov commented:
Well, using SMTP to send from Ruby fails with complaints about a self-signed certificate I can't work around in reasonable time, so I added an Aptfile installing postfix(1) and changed the code to rely on sendmail again, and hope it would work. However, I also can't seem to run the tool on toolforge manually. A call to:

````
toolforge jobs run --image tool-erex-yomi/tool-erex-yomi:latest --command=run-cron erex-yomi
````

returns without output, but also without updating **err** and **out**, so I don't know what is still going wrong.

Is there a way to get a bash prompt on the container, so I could run the tool with the ruby debugger and try to see what's going on?

This feels harder than it should be. If it won't work as a scheduled job tonight, I'll have to give up and host the tool on my personal server henceforth. I just can't afford to spend more time futzing with this. :(

david-caro commented:
But also sendmail is not under `/usr/sbin/sendmail`, but `/layers/fagiani_apt/apt/usr/sbin/sendmail`, I changed that and now the error is postfix config: 
```
Sending... sendmail: fatal: open /etc/postfix/main.cf: No such file or directory
```

That is becoming too messy as it requires you to setup postfix service too :/

In any case, I changed it to use smtp (as https://wikitech.wikimedia.org/wiki/Help:Toolforge/Email#Sending_via_SMTP recommends), and the change is small enough and it works (tested locally to my email).

david-caro commented:
Opened https://github.com/abartov/erex_yomi/pull/20

abartov commented:
Oh, thank you! I didn't know about the SMTP settings. I'm sorry I never made time to properly learn the new Toolforge environment.

I've now run the tool manually and it worked well. I hope the scheduled run tonight will go smoothly, and then we're finally done.

Thanks for all the time you spent on this, @david-caro ! I hope when I do make time to revamp my other tools, this experience would make things go smoother.

abartov commented:
@david-caro: The scheduled nightly run did not work, and did not write to **out** or **err**. :-/

abartov commented:
Okay, I installed podman(1) (didn't know that tool) and am now able to look at the toolforge runtime environment, so that's much better.  Apparently, we are now hitting an SMTP relay rejection:
```
550 Relay not permitted (Net::SMTPFatalError)
```
Any idea why?

abartov commented:
(also: why was this not written to **err**?)

david-caro commented:
> Okay, I installed podman(1) (didn't know that tool) and am now able to look at the toolforge runtime environment, so that's much better. 

Yep, you can run the image you built from anywhere, but there's a few things that will not work, like wikireplicas, sending emails, or access to NFS (as those are only available from within toolforge).

> Apparently, we are now hitting an SMTP relay rejection:
> 
> ```
> 550 Relay not permitted (Net::SMTPFatalError)
> ```
> 
> Any idea why?

That might depend, if you are running locally, you might not be allowed to send emails using toolforge (only machines from within toolforge can send emails).

Can you paste the exact command you used (and from where you used it)? That will help a lot trying to debug/reproduce.

Frow within toolforge, I was able to send myself an email using the code above (changing the destination to my email):

![image](https://github.com/abartov/erex_yomi/assets/587148/418258f0-1b35-43ec-829d-9b5a01c33404)

abartov commented:
Ah, yes, I was running locally. How can I get a shell of the container from within toolforge?

the command I ran was the one you offered:

```
podman run --rm -ti --entrypoint launcher tools-harbor.wmcloud.org/tool-erex-yomi/tool-erex-yomi:latest bash
```
And then from the shell, I ran ./erex_yomi.rb, after editing it to email only me rather than the mailing list, for testing.

david-caro commented:
It's something we are working on, as of right now you might be able to get a shell with:
```
toolforge> toolforge webservice buildservice shell
# You need the launcher bash to load the buildpack-installed environment, python libs, etc.
I have no name!@shell-1708280750:/workspace$  launcher bash
I have no name!@shell-1708280750:/workspace$
```
it will not have any NFS mounts (working on it), but everything else should work. I think that for your script, that only means that `err` and `out` are not available, but you should be able to run `./erex_yomi.rb` directly.

abartov commented:
Okay, that did work! An email was sent.
And apparently, emails were sent earlier when I ran it manually, though err and out were still not updated. So I guess we're done, though it would be nice to ensure err and out do work, so I can see problems if they arise in the future.

david-caro commented:
Happy to hear!
Yep, I'll try next week to make buildservice-based jobs behave the same as regular jobs (logging to file logs), I hope that will make the custom wrapper we had to use not needed, and make it more reliable :)
